pennylane/gradients/gradient_transform.py

```diff
@@ -411,54 +411,81 @@ def _contract_qjac_with_cjac(qjac, cjac, tape):
         cjac = cjac[0]
 
     cjac_is_tuple = isinstance(cjac, tuple)
-    if not cjac_is_tuple:
-        is_square = cjac.ndim == 2 and cjac.shape[0] == cjac.shape[1]
+    # skip_cjac = False
+    # if not cjac_is_tuple:
+    # is_square = cjac.ndim == 2 and cjac.shape[0] == cjac.shape[1]
 
-        if not qml.math.is_abstract(cjac) and (
-            is_square and qml.math.allclose(cjac, qml.numpy.eye(cjac.shape[0]))
-        ):
-            # Classical Jacobian is the identity. No classical processing is present in the QNode
-            return qjac
+    # if not qml.math.is_abstract(cjac) and (
+    # is_square and qml.math.allclose(cjac, qml.numpy.eye(cjac.shape[0]))
+    # ):
+    # skip_cjac = True
 
     multi_meas = num_measurements > 1
 
     if cjac_is_tuple:
-        multi_params = True
+        single_tape_param = False
     else:
         _qjac = qjac
         if multi_meas:
             _qjac = _qjac[0]
         if has_partitioned_shots:
             _qjac = _qjac[0]
-        multi_params = isinstance(_qjac, tuple)
+        single_tape_param = not isinstance(_qjac, tuple)
 
     tdot = partial(qml.math.tensordot, axes=[[0], [0]])
 
-    if not multi_params:
+    if single_tape_param:
         # Without dimension (e.g. expval) or with dimension (e.g. probs)
         def _reshape(x):
             return qml.math.reshape(x, (1,) if x.shape == () else (1, -1))
 
         if not (multi_meas or has_partitioned_shots):
-            # Single parameter, single measurements
+            # Single parameter, single measurements, no shot vector
+            # if skip_cjac:
+            #     return qml.math.moveaxis(_reshape(qjac), 0, -1)
             return tdot(_reshape(qjac), cjac)
 
         if not (multi_meas and has_partitioned_shots):
+            # Single parameter, multiple measurements or shot vector, but not both
+            # if skip_cjac:
+            # return tuple(qml.math.moveaxis(_reshape(q), 0, -1) for q in qjac)
             return tuple(tdot(_reshape(q), cjac) for q in qjac)
 
-        # Single parameter, multiple measurements
+        # Single parameter, multiple measurements, and shot vector
+        # if skip_cjac:
+        #     return tuple(tuple(qml.math.moveaxis(_reshape(_q), 0, -1) for _q in q) for q in qjac)
         return tuple(tuple(tdot(_reshape(_q), cjac) for _q in q) for q in qjac)
 
     if not multi_meas:
         # Multiple parameters, single measurement
         qjac = qml.math.stack(qjac)
         if not cjac_is_tuple:
+            # if skip_cjac:
+            #     return qml.math.moveaxis(qjac, 0, -1)
+            if has_partitioned_shots:
+                return tuple(tdot(qml.math.stack(q), qml.math.stack(cjac)) for q in qjac)

```

Comment:
Previously, this branch did not support the case where we have a shot vector but only a single measurement.
As we remove the special logic block above, we need to support it now, in order to not deprecate support.

Positive side effect: Shot vector+single measurement now also is supported with classical processing in the node.